### ISSUE TITLE
[feature] Add clang 12 support for conan

### ISSUE THREAD

**Issue by Paingouin** [conan-io/conan#8033]
Hi,

I use conan with cmake for a small multi-plateform project (SDL2/glew/gtest) , I'm quite happy with it, as it work without any problem so far. :)

The only problem I have is when I wanted to add an emscriptem port of the project : the version of clang used by the emscriptem SDK seems to be clang 12. 

`> emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 2.0.7
clang version 12.0.0 (Cswircachegitchromium.googlesource.com-external-github.com-llvm-llvm--project 25a8881b724abf7251a9278e72224af7e82cb9c2)
Target: x86_64-pc-windows-msvc
Thread model: posix `

As I have understood it's complicated to add a new version of a compiler to conan ( we cannot only change the conf/__init__.py  ) and I don't have the level and knowledge to do this modification. 

Can someone help me ?

Thanks for your help.

**Comment by memsharded** [comment]
Hi @Paingouin 

apple-clang 12 has already been added to Conan: https://github.com/conan-io/conan/blob/bbfb6b5ddfd2beadeffc3f4b262faec4373f2b50/conans/client/conf/__init__.py#L90

Is apple-clang the one that you want?

If it is clang, 12 has not been released yet: https://releases.llvm.org/download.html. It is added when it is released. You can definitely try adding it to your ``conf/settings.yml``, it is true that a couple of things might be missing, like the corresponding C++ standard default of the compiler, but it is definitely worth trying. Just add it to your settings.yml and let us know how it goes.

**Comment by Paingouin** [comment]
Hi @memsharded :) 

thanks for the quick answer !
It's seems that emscripten use the non-officialy release of llvm 12.

But I've tried to modify the settings.yml file and it looks like it work ! 

To be fair I don't really know why they use the unstable version of llvm the sdk instead of the stable one but I guess it's easier to maintain like this on their side.

Thanks again ! 👍

**Comment by memsharded** [comment]
Happy to know that it works @Paingouin !

Lets keep this issue open until it is officially released. Thanks.

**Comment by offa** [comment]
@memsharded Clang-12 is available officially now.

**Comment by SSE4** [comment]
apple-clang 12 is already there: https://github.com/conan-io/conan/pull/7722/files#diff-d9e6bd075cc423e1741e5ccd98478ecad42edc871876d02f63e2965c3ef07cbeR92
regular clang 12 isn't in default settings.yml yet
release notes: https://releases.llvm.org/12.0.0/tools/clang/docs/ReleaseNotes.html

**Comment by SSE4** [comment]
now has C++23 (`-std=c++2b`):
```
note: use 'c++98' or 'c++03' for 'ISO C++ 1998 with amendments' standard
note: use 'gnu++98' or 'gnu++03' for 'ISO C++ 1998 with amendments and GNU extensions' standard
note: use 'c++11' for 'ISO C++ 2011 with amendments' standard
note: use 'gnu++11' for 'ISO C++ 2011 with amendments and GNU extensions' standard
note: use 'c++14' for 'ISO C++ 2014 with amendments' standard
note: use 'gnu++14' for 'ISO C++ 2014 with amendments and GNU extensions' standard
note: use 'c++17' for 'ISO C++ 2017 with amendments' standard
note: use 'gnu++17' for 'ISO C++ 2017 with amendments and GNU extensions' standard
note: use 'c++20' for 'ISO C++ 2020 DIS' standard
note: use 'gnu++20' for 'ISO C++ 2020 DIS with GNU extensions' standard
note: use 'c++2b' for 'Working draft for ISO C++ 2023 DIS' standard
note: use 'gnu++2b' for 'Working draft for ISO C++ 2023 DIS with GNU extensions' standard
```

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/8828 merged, this will be released in next Conan 1.36

**Comment by IceflowRE** [comment]
> 
> 
> #8828 merged, this will be released in next Conan 1.36

One question:
Why does this exist in first place? Also it will brake ci's quite often when relying on the delayed conan release with this value.

**Comment by memsharded** [comment]
> One question:
Why does this exist in first place? Also it will brake ci's quite often when relying on the delayed conan release with this value

To control the valid values for settings. As they are hashed to form the ``package_id``, every minor disagreement in settings, like calling a version 12, or 12.0, or compiler Gcc vs gcc, could easily cause an unsolvable divergence of incompatible ``package_id``. 
So the definition of valid common values is very necessary. And in practice it is not a big issue. CIs doesn't break when a new compiler version is released. It needs to be installed in the CI servers first. Then, you should be relying on the default auto-detected profile, which is also not recommended practice. In production, defining your own profiles is the recommended way.
